obj: scalability improvements around run reuse

pbalcer opened the pull request:
In my synthetic benchmarks I've observed 2x improvement in the runtime state performance when throwing lots of threads at the allocator (100++).

I'd like to ask for a diligent review of `heap_run_process_bitmap_value`.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/pmem/pmdk/2884)
<!-- Reviewable:end -->

lplewa commented:
Review status: 0 of 5 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---

*[src/common/sys_util.h, line 85 at r1](https://reviewable.io/reviews/pmem/pmdk/2884#-LAx8X0V3G09S64NZvXA:-LAx8X0V3G09S64NZvXB:b-9tggbm) ([raw file](https://github.com/pmem/pmdk/blob/7ac3b1fc88280b6dfcd86d57b80a3023ec7252d2/src/common/sys_util.h#L85)):*
> ```objc
> 	if (tmp) {
> 		errno = tmp;
> 		FATAL("!util_mutex_lock");
> ```

revert or update other util_mutex functions

---


*Comments from [Reviewable](https://reviewable.io/reviews/pmem/pmdk/2884)*
<!-- Sent from Reviewable.io -->

pbalcer commented:
Review status: 0 of 7 files reviewed at latest revision, 1 unresolved discussion.

---

*[src/common/sys_util.h, line 85 at r1](https://reviewable.io/reviews/pmem/pmdk/2884#-LAx8X0V3G09S64NZvXA:-LB-zJU1D6T1apjetuTt:b-896fix) ([raw file](https://github.com/pmem/pmdk/blob/7ac3b1fc88280b6dfcd86d57b80a3023ec7252d2/src/common/sys_util.h#L85)):*
<details><summary><i>Previously, lplewa (Łukasz Plewa) wrote…</i></summary><blockquote>

revert or update other util_mutex functions
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/pmem/pmdk/2884)*
<!-- Sent from Reviewable.io -->

krzycz commented:
Reviewed 2 of 5 files at r1, 3 of 4 files at r2, 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks broke.

---



*Comments from [Reviewable](https://reviewable.io/reviews/pmem/pmdk/2884)*
<!-- Sent from Reviewable.io -->

ldorau commented:
Reviewed 1 of 5 files at r1, 2 of 4 files at r2, 2 of 2 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved.

---

*[src/common/sys_util.h, line 92 at r4](https://reviewable.io/reviews/pmem/pmdk/2884#-LCYxUGaE1qjB-QGJFd4:-LCYxUGaE1qjB-QGJFd5:b-g6mxnr) ([raw file](https://github.com/pmem/pmdk/blob/6334738fd9f952c49d4f9a689bb83fa2544b449a/src/common/sys_util.h#L92)):*
> ```objc
>  * util_mutex_trylock -- os_mutex_trylock variant that never fails from
>  * caller perspective (other than EBUSY). If util_mutex_trylock failed, this
>  * function aborts the program.
> ```

Maybe add some info about the return value? When it returns 0 and 1?

---


*Comments from [Reviewable](https://reviewable.io/reviews/pmem/pmdk/2884)*
<!-- Sent from Reviewable.io -->

marcinslusarz commented:
Review status: 5 of 9 files reviewed, 2 unresolved discussions (waiting on @krzycz and @pbalcer)

---

*[src/common/sys_util.h, line 92 at r4](https://reviewable.io/reviews/pmem/pmdk/2884#-LCYxUGaE1qjB-QGJFd4:-LFHzm_N4-waJ-tVok1y:b32v57) ([raw file](https://github.com/pmem/pmdk/blob/6334738fd9f952c49d4f9a689bb83fa2544b449a/src/common/sys_util.h#L92)):*
<details><summary><i>Previously, ldorau (Lukasz Dorau) wrote…</i></summary><blockquote>

Maybe add some info about the return value? When it returns 0 and 1?
</blockquote></details>

+1

---

*[src/libpmemobj/recycler.c, line 86 at r5](https://reviewable.io/reviews/pmem/pmdk/2884#-LFNR3csAgC41WdkG_9h:-LFNR3cu-XyHVobMZ6By:ba5hgax) ([raw file](https://github.com/pmem/pmdk/blob/af48b89b8d12277528ebd1c7dca20016bf0c437e/src/libpmemobj/recycler.c#L86)):*
> ```C
> 	 * how often should the memory block scores be recalculated.
> 	 */
> 	size_t unaccounted_units[MAX_CHUNK];
> ```

This counter will be shared by different zones. Is this intended? If it is please expand the comment above.

---


*Comments from [Reviewable](https://reviewable.io/reviews/pmem/pmdk/2884)*
<!-- Sent from Reviewable.io -->

pbalcer commented:
Review status: 5 of 9 files reviewed, 2 unresolved discussions (waiting on @krzycz, @ldorau, and @pbalcer)

---

*[src/common/sys_util.h, line 92 at r4](https://reviewable.io/reviews/pmem/pmdk/2884#-LCYxUGaE1qjB-QGJFd4:-LFXKfoM95VYI_DOgw_F:b-896fix) ([raw file](https://github.com/pmem/pmdk/blob/6334738fd9f952c49d4f9a689bb83fa2544b449a/src/common/sys_util.h#L92)):*
<details><summary><i>Previously, marcinslusarz (Marcin Ślusarz) wrote…</i></summary><blockquote>

+1
</blockquote></details>

Done.

---

*[src/libpmemobj/recycler.c, line 86 at r5](https://reviewable.io/reviews/pmem/pmdk/2884#-LFNR3csAgC41WdkG_9h:-LFXKgFQ99OymHeRgyUA:bh8lvhb) ([raw file](https://github.com/pmem/pmdk/blob/af48b89b8d12277528ebd1c7dca20016bf0c437e/src/libpmemobj/recycler.c#L86)):*
<details><summary><i>Previously, marcinslusarz (Marcin Ślusarz) wrote…</i></summary><blockquote>

This counter will be shared by different zones. Is this intended? If it is please expand the comment above.
</blockquote></details>

Yes, it is. And I know we can do this way better, which I plan to do eventually. But for now, I've tested this with terabyte sized pools with overnight random malloc/free workloads and it scales well.

---


*Comments from [Reviewable](https://reviewable.io/reviews/pmem/pmdk/2884)*
<!-- Sent from Reviewable.io -->

marcinslusarz commented on src/common/sys_util.h:
```diff
@@ -90,6 +90,7 @@ util_mutex_lock(os_mutex_t *m)
  * util_mutex_trylock -- os_mutex_trylock variant that never fails from
  * caller perspective (other than EBUSY). If util_mutex_trylock failed, this
  * function aborts the program.
+ * Returns 0 if locked successfully, otherwise returns EBUSY.
```
Doesn't belong to this patch.

marcinslusarz commented on src/common/sys_util.h:
```diff
@@ -99,7 +100,7 @@ util_mutex_trylock(os_mutex_t *m)
 		FATAL("!os_mutex_trylock");
 	}
-	return tmp == EBUSY;
+	return tmp;
```
Doesn't belong to this patch.

marcinslusarz commented on src/libpmemobj/heap.c:
```diff
@@ -466,7 +466,7 @@ heap_run_process_bitmap_value(struct bucket *b, struct memory_block *m,
 		}
 	} while (shift != BITS_PER_VALUE);
-	
+
```
Doesn't belong to this patch.

marcinslusarz commented on src/test/obj_locks/TEST1:
```diff
@@ -38,7 +38,7 @@
 . ../unittest/unittest.sh
 
-require_test_type medium
+require_test_type all
```
Doesn't belong to this patch.

marcinslusarz commented on src/test/obj_tx_locks/TEST1:
```diff
@@ -38,7 +38,7 @@
 . ../unittest/unittest.sh
 
-require_test_type medium
+require_test_type long
```
Doesn't belong to this patch.

pbalcer commented:
Review status: 6 of 8 files reviewed, 5 unresolved discussions (waiting on @janekmi, @ldorau, @pbalcer, and @marcinslusarz)

---

*[src/libpmemobj/heap.c, line 813 at r6](https://reviewable.io/reviews/pmem/pmdk/2884#-LFv14EJ90La-YfoCjij:-LFw4YM8DpraX50jcVkw:b-82q0cb) ([raw file](https://github.com/pmem/pmdk/blob/e7ee27109c245b967e83232024286e7ccf85599d/src/libpmemobj/heap.c#L813)):*
<details><summary><i>Previously, janekmi (Jan Michalski) wrote…</i></summary><blockquote>

Why do you adding here additional ```heap_run_reuse``` step?
</blockquote></details>

On current master every time we search for memory, `heap_reuse_from_recycler` will first try to find all unaccounted units, which is very expensive. This relatively small change makes it so that we will first look in the recycler if there are any *existing* free blocks that match, and only if that fails we will recalculate.

This is one of the trade-offs between fragmentation and performance, it would be better to always have up-to-date runtime tracking info in the recycler, but it would be very computationally expensive.

---

*[src/libpmemobj/recycler.c, line 365 at r6](https://reviewable.io/reviews/pmem/pmdk/2884#-LFv1pMb0pnr-DTYe5mA:-LFw5JPQEZZNkTErNi9L:bi0vuoj) ([raw file](https://github.com/pmem/pmdk/blob/e7ee27109c245b967e83232024286e7ccf85599d/src/libpmemobj/recycler.c#L365)):*
<details><summary><i>Previously, janekmi (Jan Michalski) wrote…</i></summary><blockquote>

Maybe it is a good idea to comment this line. Understanding this line took me the most time.
</blockquote></details>

This mechanism is described in recycler declaration.

---

*[src/test/obj_tx_locks/TEST1, line 41 at r6](https://reviewable.io/reviews/pmem/pmdk/2884#-LBtkjGUpp0raBtUabTI-r6-41:-LFw5yyz4yJfafQZSfpk:b-896fix) ([raw file](https://github.com/pmem/pmdk/blob/e7ee27109c245b967e83232024286e7ccf85599d/src/test/obj_tx_locks/TEST1#L41)):*
<details><summary><i>Previously, marcinslusarz (Marcin Ślusarz) wrote…</i></summary><blockquote>

Doesn't belong to this patch.
</blockquote></details>

Done.

---

*[src/test/obj_locks/TEST1, line 41 at r6](https://reviewable.io/reviews/pmem/pmdk/2884#-LBtzMI33dHQ7JY0CQND-r6-41:-LFw5ylP-S11G-hqZSHY:b-896fix) ([raw file](https://github.com/pmem/pmdk/blob/e7ee27109c245b967e83232024286e7ccf85599d/src/test/obj_locks/TEST1#L41)):*
<details><summary><i>Previously, marcinslusarz (Marcin Ślusarz) wrote…</i></summary><blockquote>

Doesn't belong to this patch.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/pmem/pmdk/2884)*
<!-- Sent from Reviewable.io -->

marcinslusarz commented:
Reviewed 1 of 4 files at r5, 5 of 6 files at r6, 1 of 3 files at r7, 1 of 1 files at r8.
Review status: 7 of 8 files reviewed, 5 unresolved discussions (waiting on @ldorau, @janekmi, @marcinslusarz, and @pbalcer)

---

*[src/test/obj_locks/TEST2, line 41 at r8](https://reviewable.io/reviews/pmem/pmdk/2884#-LG0h71J7Mojpg3591r9:-LG0h71J7Mojpg3591rA:b-cmsk5n) ([raw file](https://github.com/pmem/pmdk/blob/215c705bb803d42a6594f44c353ef4f347ac5c9b/src/test/obj_locks/TEST2#L41)):*
> ```
> . ../unittest/unittest.sh
> 
> require_test_type long
> ```

Why? And why in "obj: don't recalc runs if not necessary" patch?

---

*[src/test/obj_tx_locks/TEST1, line 41 at r6](https://reviewable.io/reviews/pmem/pmdk/2884#-LBtkjGUpp0raBtUabTI-r6-41:-LG0eLZJ6O4ix9BM6v8c:bquacbo) ([raw file](https://github.com/pmem/pmdk/blob/e7ee27109c245b967e83232024286e7ccf85599d/src/test/obj_tx_locks/TEST1#L41)):*
<details><summary><i>Previously, pbalcer (Piotr Balcer) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

It's still in "obj: don't recalc runs if not necessary" patch and I don't understand the motivation for such change.

---


*Comments from [Reviewable](https://reviewable.io/reviews/pmem/pmdk/2884)*
<!-- Sent from Reviewable.io -->

pbalcer commented:
Review status: all files reviewed, 4 unresolved discussions (waiting on @ldorau, @pbalcer, and @marcinslusarz)

---

*[src/libpmemobj/recycler.c, line 365 at r6](https://reviewable.io/reviews/pmem/pmdk/2884#-LFv1pMb0pnr-DTYe5mA:-LG4XUKz8DQ9Zi6Sddl8:bsj1rjp) ([raw file](https://github.com/pmem/pmdk/blob/e7ee27109c245b967e83232024286e7ccf85599d/src/libpmemobj/recycler.c#L365)):*
<details><summary><i>Previously, janekmi (Jan Michalski) wrote…</i></summary><blockquote>

I see but you do not describe there how do you estimate unaccounted units.
</blockquote></details>

?
I don't estimate anything. 

---

*[src/test/obj_locks/TEST2, line 41 at r8](https://reviewable.io/reviews/pmem/pmdk/2884#-LG0h71J7Mojpg3591r9:-LG4YREBA_9KiEKraAG7:bt1v6y9) ([raw file](https://github.com/pmem/pmdk/blob/215c705bb803d42a6594f44c353ef4f347ac5c9b/src/test/obj_locks/TEST2#L41)):*
<details><summary><i>Previously, marcinslusarz (Marcin Ślusarz) wrote…</i></summary><blockquote>

Why? And why in "obj: don't recalc runs if not necessary" patch?
</blockquote></details>

I removed it now.

---

*[src/test/obj_tx_locks/TEST1, line 41 at r6](https://reviewable.io/reviews/pmem/pmdk/2884#-LBtkjGUpp0raBtUabTI-r6-41:-LG4YVag1OtNl_R9EGzw:b71kxb) ([raw file](https://github.com/pmem/pmdk/blob/e7ee27109c245b967e83232024286e7ccf85599d/src/test/obj_tx_locks/TEST1#L41)):*
<details><summary><i>Previously, marcinslusarz (Marcin Ślusarz) wrote…</i></summary><blockquote>

It's still in "obj: don't recalc runs if not necessary" patch and I don't understand the motivation for such change.
</blockquote></details>

It was probably a mistake. Or, more likely, it was the same drd issue /w open.
Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/pmem/pmdk/2884)*
<!-- Sent from Reviewable.io -->

marcinslusarz commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 2 files at r9.
Review status: all files reviewed, 2 unresolved discussions (waiting on @ldorau and @janekmi)

---



*Comments from [Reviewable](https://reviewable.io/reviews/pmem/pmdk/2884#-:-LG9sN8h2A3OsARjXufG:bnfp4nl)*
<!-- Sent from Reviewable.io -->

pbalcer commented:
Review status: 5 of 6 files reviewed, 2 unresolved discussions (waiting on @janekmi, @marcinslusarz, @ldorau, and @pbalcer)

---

*[src/libpmemobj/recycler.c, line 365 at r6](https://reviewable.io/reviews/pmem/pmdk/2884#-LFv1pMb0pnr-DTYe5mA:-LG9tCgZ45snHCZMfQFi:b-j7wwf1) ([raw file](https://github.com/pmem/pmdk/blob/e7ee27109c245b967e83232024286e7ccf85599d/src/libpmemobj/recycler.c#L365)):*
<details><summary><i>Previously, pbalcer (Piotr Balcer) wrote…</i></summary><blockquote>

?
I don't estimate anything. 
</blockquote></details>

I've added a comment, and I also moved the continue to be before the sub - this might have caused worse fragmentation in later parts of the pool.

---


*Comments from [Reviewable](https://reviewable.io/reviews/pmem/pmdk/2884)*
<!-- Sent from Reviewable.io -->

janekmi commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r10.
Review status: :shipit: complete! all files reviewed, all discussions resolved

---



*Comments from [Reviewable](https://reviewable.io/reviews/pmem/pmdk/2884#-:-LGUHzz2-wzM0rqHai4n:bnfp4nl)*
<!-- Sent from Reviewable.io -->

marcinslusarz commented:
Reviewed 1 of 1 files at r10.
Review status: :shipit: complete! all files reviewed, all discussions resolved

---



*Comments from [Reviewable](https://reviewable.io/reviews/pmem/pmdk/2884)*
<!-- Sent from Reviewable.io -->
